### ISSUE TITLE
Support zero gradients in adjoint plugin

### ISSUE THREAD

**Issue by yaugenst** [flexcompute/tidy3d#1217]
## Problem description

Currently, if a monitor records amplitudes that are (close to) zero, the adjoint plugin does not place sources in the adjoint simulation (presumably due to [this line](https://github.com/flexcompute/tidy3d/blob/f60bf47f84264bc622e548fa41dfc344fb85ceaf/tidy3d/plugins/adjoint/components/data/monitor_data.py#L82)), leading to the following exception:
```
jax._src.source_info_util.JaxStackTraceBeforeTransformation: pydantic.v1.error_wrappers.ValidationError: 1 validation error for AdjointBatch
jobs -> __root__
  No sources in simulation. (type=value_error.setup)
```

This precludes the use of functions with "binary" gradients such as `jnp.min()`, `jnp.max()` etc. in adjoint-based inverse design with tidy3d, for example when optimizing the minimum transmission of a mode monitor over a batch of simulations. The current workaround is to just not let adjoint source amplitudes go to zero, either by implementing `custom_vjp`'s (and "faking it") or avoiding the use of those functions altogether.

## Proposed solution

It would be nice if tidy3d could support such "zero gradients" natively. My idea would be to piggyback on the existing check whether or not sources are placed in a simulation, and if no sources are placed in an adjoint simulation the simulation should not even attempt to run but instead return zeroed "dummy" monitors/fields. This will automatically lead to the correct (zero) gradients downstream for those monitors, since the gradient is proportional to $E_{\text{fwd}} \cdot E_{\text{adj}}$.

**Comment by momchil-flex** [comment]
I don't remember anymore why we introduced this, but I remember there was a good reason, and I think we did want to consider a better solution (e.g. along the lines of what you propose) but for some reason it wasn't super straightforward? But yeah we hadn't anticipated this specific issue.

**Comment by tylerflex** [comment]
yea I've come across this from time to time and just haven't fixed it yet. The reason we exclude sources with zero amplitudes is because if we added all of the sources, we'd have one for every element in the simulation data, which could easily be far too many. I think the proposed solution is the way to go. I just am not sure how to provide dummy fields in tidy3d. I guess one approach is to copy over the forward simulation data and just multiply all of the field data by 0 one by one.

**Comment by tylerflex** [comment]
@yaugenst I'm working on a solution in #1218. Might need to wait until our next patch or pre-release. For the meantime, you could try to use a smooth max or min function. Eg. using `jnp.softmax` to come up with a distribution of weights over your batch. https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.softmax.html might be worth considering either way for more stable optimization. Will keep you updated on this!

**Comment by yaugenst** [comment]
Awesome, thanks for looking into it! And yes you're right, I guess `logsumexp` is the canonical way to do this. Maybe it might even make sense raise a warning in the "no adjoint source" case, as something might be going on that the user wasn't intending. For example, aside from non-differentiable functions, this can also happen if sources/monitors in an adjoint simulation are set up in a way that the monitors do not register any fields (depending on the initialization of the design region). So probably good to have this functionality either way.

**Comment by tylerflex** [comment]
Yea I actually did add a warning in my [PR](https://github.com/flexcompute/tidy3d/blob/c8b630ecdf6e2872cab4299dbddf635429c10344/tidy3d/plugins/adjoint/components/data/sim_data.py#L180-L185) because of this very reason. I figure a large portion of the time, this error could pop up due to some typo in the objective function rather than intentionally.

**Comment by tylerflex** [comment]
ok this will be released in 2.5.0rc2 on Monday, closing this issue

**Comment by tylerflex** [comment]
Just FYI this was released with `2.5.0rc2` today.

**Comment by yaugenst** [comment]
Awesome, thanks!
